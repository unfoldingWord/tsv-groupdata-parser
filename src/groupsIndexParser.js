import fs from 'fs-extra';
import path from 'path-extra';
import { getGroupName } from './helpers/resourcesHelpers';

/**
 * make sure this groupId is in a category
 * @param {String} groupId
 * @param {String} groupName - localized group name
 * @param {Object} categorizedGroupsIndex
 * @param {String} categoryName - category that should contain this groupId
 */
const addGroupToCategory = (groupId, groupName, categorizedGroupsIndex, categoryName) => {
  const groupIndexItem = getGroupIndex(groupId, groupName);

  // Only add the groupIndexItem if it isn't already in the category's groups index.
  if (!categorizedGroupsIndex[categoryName].some(e => e.id === groupIndexItem.id)) {
    categorizedGroupsIndex[categoryName].push(groupIndexItem); // adding group Index Item
  }
};

export const generateGroupsIndex = (tnCategoriesPath, taCategoriesPath) => {
  const categorizedGroupsIndex = {
    discourse: [],
    numbers: [],
    figures: [],
    culture: [],
    grammar: [],
    other: [],
  };
  const errors = [];

  const isDirectory = item => fs.lstatSync(path.join(tnCategoriesPath, item)).isDirectory();
  const categories = fs.readdirSync(tnCategoriesPath).filter(isDirectory);

  categories.forEach(categoryName => {
    const booksPath = path.join(tnCategoriesPath, categoryName, 'groups');
    const books = fs.readdirSync(booksPath).filter(item => item !== '.DS_Store');

    books.forEach(bookid => {
      const groupDataPath = path.join(booksPath, bookid);
      const groupDataFiles = fs.readdirSync(groupDataPath).filter(filename => path.extname(filename) === '.json');
      let taArticleCategory;
      let groupName;
      let contextId;

      groupDataFiles.forEach(groupDataFile => {
        try {
          const filePath = path.join(groupDataPath, groupDataFile);
          const groupId = groupDataFile.replace('.json', '');
          const groupData = fs.readJsonSync(filePath);
          taArticleCategory = null;
          groupName = null;
          let foundLocalization = false;
          contextId = null;

          if (groupData.length > 0) {
            for (let i = 0; i < groupData.length; i++ ) {
              const contextId = groupData[i].contextId;
              taArticleCategory = getArticleCategory(contextId.occurrenceNote, groupId);

              try {
                if (!taArticleCategory) {
                  throw new Error(`Link in Occurrence Note '${contextId.occurrenceNote}' does not have category for check at contextID: ${JSON.stringify(contextId)}`);
                }

                const fileName = groupId + '.md';
                const articlePath = path.join(taCategoriesPath, taArticleCategory, fileName);
                groupName = getGroupName(articlePath);
                addGroupToCategory(groupId, groupName, categorizedGroupsIndex, categoryName);
                foundLocalization = true;
                break; // we got the category, so don't need to search anymore
              } catch (e) {
                let message = `error finding group name: groupId: '${groupId}', index: '${i}' in bookId '${bookid}, taArticleCategory: ${taArticleCategory}, contextID: ${JSON.stringify(contextId)}' `;
                console.error('generateGroupsIndex() - ' + message, e);
              }
            }

            if (!foundLocalization) {
              addGroupToCategory(groupId, groupId, categorizedGroupsIndex, categoryName); // add entry even though we could not find localized description
              console.error(`Could not find localization for '${groupId}' in '${categoryName}', adding stub entry`);
            }
          }
        } catch (e) {
          let message = `error processing entry: bookid: ${bookid}, groupDataFile: ${groupDataFile}, taArticleCategory: ${taArticleCategory}, groupName: ${groupName}, contextID: ${JSON.stringify(contextId)}: `;
          console.error('generateGroupsIndex() - ' + message, e);
          errors.push(message + e.toString());
        }
      });
    });
  });

  if (errors.length) {
    throw new Error(`generateGroupsIndex() - error processing index:\n${errors.join('\n')}`);
  }

  return categorizedGroupsIndex;
};

/**
 * Gets the category of the given groupId from the links in the occurrenceNote
 * @param {string} occurrenceNote
 * @param {string} groupId
 * @returns {string} the matched category
 */
export const getArticleCategory = (occurrenceNote, groupId) => {
  if (occurrenceNote && groupId) {
    const pattern = '(?<=rc:\\/\\/[^\\/]+\\/ta\\/man\\/)[^\\/]+?(?=\\/' + groupId + ')';
    const categoryRE = new RegExp(pattern);
    const match = occurrenceNote.match(categoryRE);

    if (match) {
      return match[0];
    }
  }
  return null;
};

const getGroupIndex = (groupId, groupName) => ({
  id: groupId,
  name: groupName,
});
