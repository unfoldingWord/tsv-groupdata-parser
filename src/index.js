import ManageResource from './helpers/ManageResourceAPI';
import * as verseHelpers from './helpers/verseHelpers';
import * as wordOccurrenceHelpers from './helpers/wordOccurrenceHelpers';
export {
  ManageResource,
  verseHelpers,
  wordOccurrenceHelpers,
};
export {
  cleanGroupId,
  generateGroupDataItem,
  tsvToGroupData,
  tnJsonToGroupData,
} from './tsvToGroupData';
export { convertTsv9to7 } from './tsvConversion';
export { generateGroupsIndex } from './groupsIndexParser';
export {
  getCategoryForGroupId, categorizeGroupData, tNotesCategories,
} from './tNoteGroupIdCategorization';
export { formatAndSaveGroupData, saveGroupsIndex } from './helpers/filesystemHelpers';
