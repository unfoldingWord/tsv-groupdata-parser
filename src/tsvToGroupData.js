/* eslint-disable no-useless-escape */
import path from 'path-extra';
import fs from 'fs-extra';
import tsvtojson from 'tsvtojson';
import { categorizeGroupData } from './tNoteGroupIdCategorization';
import ManageResource from './helpers/ManageResourceAPI';
import { getWordOccurrencesForQuote } from './helpers/wordOccurrenceHelpers';
import { ELLIPSIS } from './utils/constants';
import {
  translationHelps, getLatestVersionInPath, getGroupName, getBibleIdForLanguage,
} from './helpers/resourcesHelpers';
import { hasEllipsis } from './helpers/ellipsisHelpers';

/**
 * Parses a book tN TSVs and returns an object holding the lists of group ids.
 * @param {string} filepath path to tsv file.
 * @param {string} toolName tC tool name.
 * @param {object} params When it includes { categorized: true }
 * then it returns the object organized by tn article category.
 * @param {string} originalBiblePath path to original bible.
 * e.g. /resources/el-x-koine/bibles/ugnt/v0.11
 * @param {string} resourcesPath path to the resources dir
 * e.g. /User/john/translationCore/resources
 * @param {string} langId
 * @returns an object with the lists of group ids which each includes an array of groupsdata.
 */
export const tsvToGroupData = async (filepath, toolName, params = {}, originalBiblePath, resourcesPath, langId) => {
  const groupData = {};
  const tsvObjects = await tsvtojson(filepath);
  const { Book: bookId } = tsvObjects[0] || {};

  try {
    const resourceApi = new ManageResource(originalBiblePath, bookId.toLowerCase());

    for (const tsvItem of tsvObjects) {
      if (tsvItem.SupportReference && tsvItem.OrigQuote) {
        const supportReference = cleanGroupId(tsvItem.SupportReference);

        if (!supportReference) {
          continue; // skip over if cleaned support reference is empty
        }
        tsvItem.OccurrenceNote = cleanOccurrenceNoteLinks(tsvItem.OccurrenceNote || '', resourcesPath, langId, bookId.toLowerCase(), tsvItem.Chapter);

        if (!tsvItem.OccurrenceNote) {
          console.warn('tsvToGroupData() - error processing item:', JSON.stringify(tsvItem));
          continue;
        }

        const chapter = parseInt(tsvItem.Chapter, 10);
        const verse = parseInt(tsvItem.Verse, 10);
        let verseString = null;

        try {
          verseString = resourceApi.getVerseString(chapter, verse);
        } catch (e) {
          console.warn(`tsvToGroupData() - error getting verse string: chapter ${chapter}, verse ${verse} from ${JSON.stringify(tsvItem)}`, e);
        }

        if (verseString) {
          if (groupData[supportReference]) {
            groupData[supportReference].push(generateGroupDataItem(tsvItem, toolName, verseString));
          } else {
            groupData[supportReference] = [generateGroupDataItem(tsvItem, toolName, verseString)];
          }
        }
      }
    }

    return params && params.categorized ? categorizeGroupData(groupData) : groupData;
  } catch (e) {
    console.error(`tsvToGroupData() - error processing filepath: ${filepath}`, e);
    throw e;
  }
};

/**
 * Cleans an incorrectly formatted group id.
 * @param {string} groupId group id string that was possibly incorrectly formatted.
 * @returns {string} correctly formatted group id.
 */
export const cleanGroupId = groupId => {
  try {
    // Make sure we only have the element at the very end of a path of /'s or :'s
    // Ex: translate:writing_background => writing_background
    const elements = groupId.split(/[/:]/);
    let cleanedId = elements[elements.length - 1];
    // Replace _ with - in groupId
    // Ex: writing_background => writing-background
    cleanedId = cleanedId.replaceAll('_', '-');
    cleanedId = cleanedId.replaceAll('<br>', '').trim(); // remove white space including html new lines
    return cleanedId;
  } catch (e) {
    console.error(`cleanGroupId() - groupId: ${groupId}`, e);
    return null;
  }
};

/**
 * Converts [[rc://lang/(ta|tw)/...]] links to a markdown link if we can find their article file locally to get the title
 * @param {string} tHelpsLink
 * @param {string} resourcesPath
 * @param {string} langId
 */
export const convertLinkToMarkdownLink = (tHelpsLink, resourcesPath, langId) => {
  try {
    const tHelpsPattern = /\[\[(rc:\/\/[\w-]+\/(ta|tn|tw)\/[^\/]+\/([^\]]+)\/([^\]]+))\]\]/g;
    const parts = tHelpsPattern.exec(tHelpsLink);
    parts.shift();
    const [rcLink, resource, category, file] = parts;
    let resourcePath = path.join(resourcesPath, langId, 'translationHelps', translationHelps[resource]);
    resourcePath = getLatestVersionInPath(resourcePath);
    let articlePath;

    if (resource === 'ta') {
      articlePath = path.join(resourcePath, category, file) + '.md';
    }

    if (resource === 'tw') {
      articlePath = path.join(resourcePath, category.split('/')[1], 'articles', file) + '.md';
    }

    if (articlePath && fs.existsSync(articlePath)) {
      const groupName = getGroupName(articlePath);

      if (groupName) {
        return '[' + groupName + '](' + rcLink + ')';
      }
    }
    return tHelpsLink;
  } catch (e) {
    console.error(`convertLinkToMarkdownLink() - invalid link: ${tHelpsLink}, resourcesPath: ${resourcesPath}, langId: ${langId}`, e);
    return null;
  }
};

/**
 * Fixes Bible links by putting in a rc:// link for the given language, book, chapter and verse
 * @param {string} link
 * @param {string} resourcesPath
 * @param {string} langId
 * @param {string} bookId
 * @param {string|int} chapter
 * @returns {string}
 */
export const fixBibleLink = (link, resourcesPath, langId, bookId, chapter) => {
  const bibleId = getBibleIdForLanguage(path.join(resourcesPath, langId, 'bibles'));

  if (! bibleId) {
    return link;
  }

  // bibleLinkPattern can match the following:
  // [Titus 2:1](../02/01.md)
  // [John 4:2](../../jhn/04/02.md] (parts[4] will be "jhn")
  // [Revelation 10:10](../10/10)
  // [Ephesians 4:1](04/01)
  // [1 Corinthians 15:12](./12.md)
  const bibleLinkPattern = /(\[[^[\]]+\])\s*\((\.+\/)*(([\w-]+)\/){0,1}?((\d+)\/)?(\d+)(\.md)*\)/g;
  const parts = bibleLinkPattern.exec(link);

  if (!parts) {
    return link;
  }
  // Example of parts indexes if the link was [1 John 2:1](../../1jn/02/01.md):
  // 0: "[1 John 2:1](../../1jn/02/01.md)"
  // 1: "[1 John 2:1]"
  // 2: "../"
  // 3: "1jn/""
  // 4: "1jn"
  // 5: "02/"
  // 6: "02"
  // 7: "01"
  // 8: ".md"

  // If the bible link is in the form (../../<bookId>/<chapter>/<verse>) we use the bookId in the link instead of the
  // one passed to this function
  let linkBookId = bookId;

  if (parts[4]) {
    linkBookId = parts[4];
  }

  // If the bible link is in the form (../<chapter>/<verse>) we use the chapter in the link instead of the one passed
  // to this function
  let linkChapter = '' + chapter; // make sure it is a string

  if (parts[6]) {
    linkChapter = parts[6];
  }
  linkChapter = linkChapter.padStart(linkBookId === 'psa' ? 3 : 2, '0'); // left pad with zeros, 2 if not Psalms, 3 if so
  return parts[1] + '(rc://' + [langId, bibleId, 'book', linkBookId, linkChapter, parts[7]].join('/') + ')';
};

/**
 * Cleans up all the links in an occurrenceNote
 * @param {string} occurrenceNote occurrence Note.
 * @param {string} resourcesPath path to the translationHelps directory that contains tA and tW article dirs
 * @param {string} langId
 * @param {string} bookId id of the book being processed
 * @param {string|int} chapter chapter of the note
 * @returns {string} occurrenceNote with clean/fixed tA links.
 */
export const cleanOccurrenceNoteLinks = (occurrenceNote, resourcesPath, langId, bookId, chapter) => {
  try {
    // Change colons in the path part of the link to a slash
    // Ex: [[rc://en/man/ta:translate:figs-activepassive]] =>
    //     [[rc://en/man/ta/translate/figs-activepassive]]
    const colonInPathPattern = /(?<=\[\[rc:[^\]]+):(?=[^\]]+\]\])/g;
    let cleanNote = occurrenceNote.replace(colonInPathPattern, '/');
    // Remove spaces between the link and right paren
    // Ex: (See: [[rc://en/man/ta:translate:figs-activepassive]] ) =>
    //     (See: [[rc://en/man/ta/translate/figs-activepassive]])
    const spaceBetweenLinkAndParenPattern = /(?<=\[\[rc:[^\]]+]]) +\)/g;
    cleanNote = cleanNote.replace(spaceBetweenLinkAndParenPattern, ')');
    // Remove invalid paren and spaces at end of the link
    // Ex: [[rc://en/man/ta:translate:figs-activepassive )]] =>
    //     [[rc://en/man/ta/translate/figs-activepassive]]
    const invalidParenInLinkPattern = /(?<=\[\[rc:[^\] )]+)[ \)]+(?=]])/g;
    cleanNote = cleanNote.replace(invalidParenInLinkPattern, ')');
    // Removes a right paren if it appears after one link and then there is another link
    // Ex: (See: [[rc://en/ta/man/translate/figs-activepassive]]) and [[rc://en/ta/man/translate/figs-idiom)]]) =>
    //     (See: [[rc://en/ta/man/translate/figs-activepassive]] and [[rc://en/ta/man/translate/figs-idiom)]])
    const rightParenBetweenLinksPattern = /(?<=\[\[rc:\/\/[^\]]+]])\)(?=[^\(]+rc:)/g;
    cleanNote = cleanNote.replace(rightParenBetweenLinksPattern, '');
    // Run cleanGroupId on the last item of the path, the groupId
    // Ex: [[rc://en/man/ta/translate/figs_activepassive]] =>
    //     [[rc://en/man/ta/translate/figs-activepassive]]
    const groupIdPattern = /(?<=\[\[rc:\/\/[^\]]+\/)[\w-]+(?=\]\])/g;
    cleanNote = cleanNote.replace(groupIdPattern, cleanGroupId);

    // Run convertLinkToMarkdownLink on each link to get their (title)[rc://...] representation
    // Ex: [[rc://en/ta/man/translate/figs_activepassive]] =>
    //     [Active or Passive](rc://en/man/ta/translate/figs_activepassive)
    if (resourcesPath && langId) {
      const tHelpsPattern = /(\[\[rc:\/\/[\w-]+\/(ta|tw)\/[^\/]+\/[^\]]+\]\])/g;

      cleanNote = cleanNote.replace(tHelpsPattern, link => {
        let convertedLink = convertLinkToMarkdownLink(link, resourcesPath, langId);

        if (!convertedLink) {
          throw new Error(`cleanOccurrenceNoteLinks() - error converting link: ${link}`);
        }
        return convertedLink;
      });
    }

    // Run fixBibleLink on each link to get a proper Bible rc link with Markdown syntax
    // Ex: [Titus 2:1](../02/01.md) => [Titus 2:1](rc://lang/ult/book/tit/02/01)
    //     [Romans 1:1](./01.md) => [Romans 1:1](rc://lang/ult/book/rom/01/01)
    //     [1 Corinthians 15:12](../../../1co/15/12) => [1 Corinthians 15:12](rc://lang/ult/book/1co/15/12)
    if (bookId && langId && resourcesPath) {
      const bibleLinkPattern = /(\[[^[\]]+\])\s*\((\.+\/)*(([\w-]+)\/){0,1}?((\d+)\/)?(\d+)(\.md)*\)/g;
      cleanNote = cleanNote.replace(bibleLinkPattern, link => fixBibleLink(link, resourcesPath, langId, bookId, chapter));
    }
    return cleanNote;
  } catch (e) {
    console.error(`cleanOccurrenceNoteLinks() - occurrenceNote: ${occurrenceNote}, resourcesPath: ${resourcesPath}, langId: ${langId}, bookId: ${bookId}, chapter: ${chapter}`, e);
    return null;
  }
};

/**
 * Returns the formatted groupData item for a given tsv item.
 * @param {object} tsvItem tsv item.
 * @param {string} toolName tool name.
 * @returns {object} groupData item.
 */
const generateGroupDataItem = (tsvItem, toolName, verseString) => {
  const { OrigQuote = '' } = tsvItem;
  // if quote has more than one word get word occurrences
  const wordOccurrencesForQuote = getWordOccurrencesForQuote(OrigQuote, verseString, true); // uses tokenizer to get list of words handle various punctuation and spacing chars
  const quote = wordOccurrencesForQuote.length > 1 || hasEllipsis(OrigQuote) ? wordOccurrencesForQuote : OrigQuote; // only use array if more than one word found
  const quoteString = OrigQuote.trim().replace(/\.../gi, ELLIPSIS);

  return {
    comments: false,
    reminders: false,
    selections: false,
    verseEdits: false,
    nothingToSelect: false,
    contextId: {
      checkId: tsvItem.ID || '',
      occurrenceNote: tsvItem.OccurrenceNote || '',
      reference: {
        bookId: tsvItem.Book.toLowerCase() || '',
        chapter: parseInt(tsvItem.Chapter, 10) || '',
        verse: parseInt(tsvItem.Verse, 10) || '',
      },
      tool: toolName || '',
      groupId: tsvItem.SupportReference || '',
      quote,
      quoteString,
      glQuote: tsvItem.GLQuote || '',
      occurrence: parseInt(tsvItem.Occurrence, 10) || 1,
    },
  };
};
