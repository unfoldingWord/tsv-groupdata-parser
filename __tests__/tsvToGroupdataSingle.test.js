import path from 'path-extra';
import fs from 'fs-extra';
// helpers
import {tsvToGroupData, tsvToGroupData7Cols} from '../src/tsvToGroupData';

// constants
const RESOURCES_PATH = path.join(__dirname, 'fixtures', 'resources');
const ORIGINAL_BIBLE_PATH_V0_11 = path.join(RESOURCES_PATH, 'el-x-koine', 'bibles', 'ugnt', 'v0.11');
const TSV_PATH_TIT = path.join(RESOURCES_PATH, '__tests__/fixtures/tsv/en_tn_57-TIT.tsv');
const RESOURCES_COPY_PATH = path.join(RESOURCES_PATH);

const TSV9_TEMPLATE = 'Book\tChapter\tVerse\tID\tSupportReference\tOrigQuote\tOccurrence\tGLQuote\tOccurrenceNote\n' +
  '{Book}\t{Chapter}\t{Verse}\t{ID}\t{SupportReference}\t{OrigQuote}\t{Occurrence}\t{GLQuote}\t{OccurrenceNote}\n';
const TSV9_TEST_DATA_DEFAULTS_TIT = {
  'Book': 'TIT',
  'Chapter': '1',
  'Verse': '15',
  'ID': 'n3wk',
  'SupportReference': 'figs-metaphor',
  'OrigQuote': 'τοῖς…μεμιαμμένοις καὶ ἀπίστοις, οὐδὲν καθαρόν',
  'Occurrence': '1',
  'GLQuote': 'to those who are corrupt and unbelieving, nothing is pure',
  'OccurrenceNote': 'Paul speaks of sinners as if they were physically dirty. Alternate translation: “if people are morally defiled and do not believe, they cannot do anything pure” or “when people are full of sin and unbelief, nothing that they do is acceptable to God” (See: [Metaphor](rc://en/ta/man/translate/figs-metaphor))',
};

const TIT_TEST = {
  ORIGINAL_BIBLE_PATH: ORIGINAL_BIBLE_PATH_V0_11,
  TSV_PATH: TSV_PATH_TIT,
  TSV_TEMPLATE: TSV9_TEMPLATE,
  TSV_TEST_DATA_DEFAULTS: TSV9_TEST_DATA_DEFAULTS_TIT,
};

// originalVerse tit 1:15 = 'πάντα καθαρὰ τοῖς καθαροῖς; τοῖς δὲ μεμιαμμένοις καὶ ἀπίστοις, οὐδὲν καθαρόν; ἀλλὰ μεμίανται αὐτῶν καὶ ὁ νοῦς, καὶ ἡ συνείδησις. \n';


const ORIGINAL_BIBLE_PATH_V0_30 = path.join(RESOURCES_PATH, 'el-x-koine', 'bibles', 'ugnt', 'v0.30_Door43-Catalog');
const TSV_PATH_PHP = path.join(RESOURCES_PATH, '__tests__/fixtures/tsv/en_tn_PHP-verse-range.tsv');

// php 1:1-2;2:5 =
// "Παῦλος καὶ Τιμόθεος, δοῦλοι Χριστοῦ Ἰησοῦ; πᾶσιν τοῖς ἁγίοις ἐν Χριστῷ Ἰησοῦ τοῖς οὖσιν ἐν Φιλίπποις, σὺν ἐπισκόποις καὶ διακόνοις: \n" +
// " χάρις ὑμῖν καὶ εἰρήνη ἀπὸ Θεοῦ Πατρὸς ἡμῶν καὶ Κυρίου Ἰησοῦ Χριστοῦ. \n" +
// " τοῦτο φρονεῖτε ἐν ὑμῖν, ὃ καὶ ἐν Χριστῷ Ἰησοῦ; "

const TSV7_TEMPLATE = 'Reference\tID\tTags\tSupportReference\tQuote\tOccurrence\tNote\n' +
  '{Reference}\t{ID}\t{Tags}\t{SupportReference}\t{OrigQuote}\t{Occurrence}\t{Note}\n';
const TSV7_TEST_DATA_DEFAULTS_PHP = {
  'Reference': '1:1-2,2:5',
  'ID': 'ezsy',
  'Tags': '',
  'SupportReference': 'rc://*/ta/man/translate/figs-abstractnouns',
  'OrigQuote': 'καὶ & ἐν',
  'Occurrence': '-1',
  'Note': ' Testing Highlighting',
};

const PHP_TEST = {
  ORIGINAL_BIBLE_PATH: ORIGINAL_BIBLE_PATH_V0_30,
  TSV_PATH: TSV_PATH_PHP,
  TSV_TEMPLATE: TSV7_TEMPLATE,
  TSV_TEST_DATA_DEFAULTS: TSV7_TEST_DATA_DEFAULTS_PHP,
  isTSV7: true,
  bookId: 'PHP',
  languageId: 'en',
};

const tests = [
  { name: 'simple ampersand with verse range 4', data: { Reference: '1:1-2,2:5', OrigQuote: 'ἡμῶν & Χριστοῦ' }, config: PHP_TEST, expect: 'quote' },
  { name: 'simple ampersand with verse range 3', data: { Reference: '1:1-2,2:5', OrigQuote: 'Χριστῷ & Ἰησοῦ' }, config: PHP_TEST, expect: 'quote' },
  { name: 'simple ampersand with verse range 2', data: { Reference: '1:1-2,2:5', OrigQuote: 'δοῦλοι & Ἰησοῦ' }, config: PHP_TEST, expect: 'quote' },
  { name: 'simple ampersand with verse range', data: { Reference: '1:1-2,2:5', OrigQuote: 'καὶ & ἐν' }, config: PHP_TEST, expect: 'quote' },
  { name: 'simple ampersand 3', data: { Reference: '1:1', OrigQuote: 'καὶ & ἐν' }, config: PHP_TEST, expect: 'quote' },
  { name: 'simple ampersand', data: { OrigQuote: 'καθαροῖς & τοῖς' }, config: TIT_TEST, expect: 'quote' },
  { name: 'simple ampersand 2', data: { OrigQuote: 'τοῖς & καὶ' }, config: TIT_TEST, expect: 'quote' },
  { name: 'all occurrences', data: { OrigQuote: 'καὶ', Occurrence: '-1' }, config: TIT_TEST, expect: 'quote' },
  { name: 'multiple ellipsis and phrases', data: { OrigQuote: 'τοῖς καθαροῖς…καὶ ἀπίστοις…τοῖς καθαροῖς…καὶ ἀπίστοις' }, config: TIT_TEST, expect: 'quote' },
  { name: 'multiple ellipsis', data: { OrigQuote: 'τοῖς…καὶ…τοῖς…καὶ' }, config: TIT_TEST, expect: 'quote' },
  { name: 'multiple ellipsis 2', data: { OrigQuote: 'τοῖς…καὶ…τοῖς…καὶ…τοῖς…καὶ' }, config: TIT_TEST, expect: 'quote' },
  { name: 'multiple ellipsis 3', data: { OrigQuote: 'τοῖς…καὶ…τοῖς…καὶ…τοῖς…καὶ…τοῖς…καὶ' }, config: TIT_TEST, expect: 'quote' },
  { name: 'simple ellipsis', data: { OrigQuote: 'τοῖς…καὶ' }, config: TIT_TEST, expect: 'quote' },
  { name: 'second occurrence', data: { OrigQuote: 'καὶ', Occurrence: '2' }, config: TIT_TEST, expect: 'quote' },
  { name: 'fifth occurrence', data: { OrigQuote: 'καὶ', Occurrence: '5' }, config: TIT_TEST, expect: 'quote' },
];

describe('tsvToGroupData() - testing quotes', () => {
  beforeEach(() => {
    fs.__resetMockFS();
    fs.__loadDirIntoMockFs(RESOURCES_COPY_PATH, RESOURCES_COPY_PATH);
  });

  for (const test_ of tests) {
    test(`Test ${test_.name}`, async () => {
      const filepath = test_.config.TSV_PATH;
      const newData = generateTestTsv(filepath, test_);
      const toolName = 'translationNotes';
      let result, supportReference;

      if (test_.config.isTSV7) {
        const params = { categorized: false };
        result = await tsvToGroupData7Cols(filepath, test_.config.bookId, RESOURCES_PATH, test_.config.languageId, toolName, test_.config.ORIGINAL_BIBLE_PATH, params);
        supportReference = newData.SupportReference.split('/')[6];
      } else {
        result = await tsvToGroupData(filepath, toolName, null, test_.config.ORIGINAL_BIBLE_PATH, RESOURCES_PATH, 'en');
        supportReference = newData.SupportReference;
      }

      const category = result[supportReference];
      const check = category[0];
      const contextId = check.contextId;
      const testValue = contextId[test_.expect];
      expect(testValue).toMatchSnapshot();
    });
  }
});

/**
 * creates a TSV file using defaults
 * @param {string} tsvPath - path for file
 * @param {object} testConfig - test configuration to use
 */
function generateTestTsv(tsvPath, testConfig) {
  const newData = {
    ...testConfig.config.TSV_TEST_DATA_DEFAULTS,
    ...testConfig.data,
  };
  let tsvContents = testConfig.config.TSV_TEMPLATE;

  for (const key of Object.keys(newData)) {
    const stringToReplace = '{' + key + '}';
    tsvContents = tsvContents.replaceAll(stringToReplace, newData[key]);
  }
  fs.writeFileSync(tsvPath, tsvContents, 'utf8');
  // const verify = fs.readFileSync(tsvPath, 'utf8');
  // console.log(verify);
  return newData;
}
