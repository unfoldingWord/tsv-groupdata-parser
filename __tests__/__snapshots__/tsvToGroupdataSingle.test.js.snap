// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tsvToGroupData() - testing quotes Test all occurrences 1`] = `"καὶ"`;

exports[`tsvToGroupData() - testing quotes Test fifth occurrence 1`] = `"καὶ"`;

exports[`tsvToGroupData() - testing quotes Test multiple ellipsis 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "καὶ",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test multiple ellipsis 2 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 3,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 3,
    "word": "καὶ",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test multiple ellipsis 3 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 3,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 3,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 4,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 4,
    "word": "καὶ",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test multiple ellipsis and phrases 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "τοῖς",
  },
  Object {
    "occurrence": 1,
    "word": "καθαροῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
  Object {
    "occurrence": 1,
    "word": "ἀπίστοις",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 3,
    "word": "τοῖς",
  },
  Object {
    "occurrence": 2,
    "word": "καθαροῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "καὶ",
  },
  Object {
    "occurrence": 2,
    "word": "ἀπίστοις",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test second occurrence 1`] = `"καὶ"`;

exports[`tsvToGroupData() - testing quotes Test simple ampersand 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "καθαροῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "τοῖς",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test simple ampersand 2 1`] = `
Array [
  Object {
    "occurrence": 2,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test simple ampersand 3 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "ἐν",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test simple ampersand with verse range 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 3,
    "word": "ἐν",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test simple ampersand with verse range 2 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "δοῦλοι",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "Ἰησοῦ",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test simple ampersand with verse range 3 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "Χριστῷ",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "Ἰησοῦ",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test simple ampersand with verse range 4 1`] = `
Array [
  Object {
    "occurrence": 1,
    "word": "ἡμῶν",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 2,
    "word": "Χριστοῦ",
  },
]
`;

exports[`tsvToGroupData() - testing quotes Test simple ellipsis 1`] = `
Array [
  Object {
    "occurrence": 2,
    "word": "τοῖς",
  },
  Object {
    "word": "…",
  },
  Object {
    "occurrence": 1,
    "word": "καὶ",
  },
]
`;
